WebXR Support

Maksims opened the pull request:
This PR removes VR APIs from engine, and adds XR support.
Currently it implements only starting/ending XR sessions (VR/AR). Input devices and reference space features will be added in another PR.

New APIs:
```javascript
// pc.XrManager as 'xr' on pc.Application
app.xr // access to XrManager
// properties
app.xr.supported // true if WebXR is supported
app.xr.active // true if WebXR Session is active
app.xr.type // null or string (pc.XRTYPE_*) of current active session
// methods
app.xr.isAvailable(type) // check if session of pc.XRTYPE_* is available on current platform
app.xr.start(camera, type, [callback]) // attempt to start XR session for CameraComponent and of pc.XRTYPE_*. With optional callback that can return Error if failed to start.
app.xr.end([callback]) // attempt to end XR session. With optional callback with Error if failed to end.
// events
app.xr.on('available', function (type, available) { } ) // event fired when availability of specified type is changed. Where type is one of pc.XRTYPE_* and available is a boolean.
app.xr.on('available:[type]', function (available) { } ) // event fired when availability of specified type is changed. Where [type] is one of pc.XRTYPE_* and available is a boolean.
app.xr.on('start', function () { } ) // event fired when session is started
app.xr.on('end', function () { } ) // event fired when session is ended

// pc.CameraComponent
entity.camera.startXr(type, [callback]) // start session from specified camera component, where type is one of pc.XRTYPE_*. With optional callback that can return Error if failed to start.
entity.camera.endXr([callback]) // attempt to end XR session. With optional callback with Error if failed to end.

// pc
pc.XRTYPE_INLINE // 'inline' - type of session which is always available
pc.XRTYPE_VR // 'immersive-vr' - type of session, offers best VR experience with access to most hardware sensors
pc.XRTYPE_AR // 'immersive-ar' - type of session, available on platforms with camera.
```

Fixes #1748

I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

Maksims commented:
Some progress here, added ability to choose session type.
Fixed issues with switching between rendering targets (backbuffer of framebuffer for XR).
Fixed issues with canvas resizing.

Tested AR on Android with chrome://flags/#webxr enabled. Surprised to see how well it works already.
For testing, can use this project: https://playcanvas.com/project/660566/overview/webxr

Maksims commented:
This PR already implements what WebVR was providing - starting/ending XR sessions.
It does not implements XR input devices, nor reference space features. This will be added in another PR.

So this can be merged, as it will provide ability to start experimenting with VR/AR.
It can be tested on this project: https://playcanvas.com/project/660566/overview/webxr

willeastcott commented:
Great stuff!

The unit tests are failing. Can you fix please?

willeastcott commented on src/framework/application.js:
```diff
@@ -1565,29 +1577,6 @@ Object.assign(pc, function () {
-                this.vr = null;
-            }
-        },
-
```
For any API that you remove, can you add stub functions that print a debug warning to src\backwards-compatibility.js please?

Maksims replied:
Added some warnings.

Maksims commented:
Added deprecation warnings, and fixed tests.

willeastcott commented:
Excellent. Can you also update this please:

https://github.com/playcanvas/engine/blob/master/examples/graphics/virtual-reality.html

Maybe even consider creating a new subfolder 'xr' and move it there (we'll probably add an AR and other related examples too!). 😄

willeastcott commented:
A few lint errors. You can auto-fix most/all:

`npm run lint -- --fix`

Maksims commented:
Fixed lint and added two examples for basic VR and basic AR.
Also updated [Starter Kit: VR](https://playcanvas.com/project/435780/overview/starter-kit-vr) project in a **webxr** branch, so can be merged to **master** once engine is deployed with this PR.

willeastcott commented on src/framework/application.js:
```diff
@@ -1808,20 +1803,27 @@ Object.assign(pc, function () {
             app._fillFrameStats(now, dt, ms);
             // #endif
 
+            if (frame) app.xr.calculateViews(frame);
```
Why not put this inside the `if (frame)` condition below?

willeastcott commented on src/framework/components/camera/component.js:
```diff
@@ -134,45 +134,26 @@ Object.assign(pc, function () {
-     * @type {pc.GraphNode}
-     * @description Queries the camera's GraphNode. Can be used to get position and rotation.
+     * @name pc.CameraComponent#xr
+     * @type Boolean
```
Boolean -> boolean

willeastcott commented on src/framework/components/camera/component.js:
```diff
@@ -134,45 +134,26 @@ Object.assign(pc, function () {
-     * @name pc.CameraComponent#node
-     * @type {pc.GraphNode}
-     * @description Queries the camera's GraphNode. Can be used to get position and rotation.
+     * @name pc.CameraComponent#xr
```
Do you think `pc.CameraComponent#xr` will be confused with `pc.Application#xr`? In other words, might the developer assume it's another `pc.XrManager` instance? Alternative naming suggestions: `xrActive`, `xrEnabled`.

Maksims replied:
Based on other naming conventions, do you think `isXr` is reasonable naming for this property?

willeastcott replied:
Just chatted with the team. The most preferred naming is `pc.CameraComponent#xrEnabled`. Let's go with that. 😄

Maksims replied:
I actually couldn't think of use case for this property, so removed it entirely.

Maksims commented:
Hm, noticed one issue with Box Projected CubeMaps, they are not projected correctly in VR.

willeastcott commented:
Interesting. Screenshot?

Maksims commented:
> Interesting. Screenshot?

![box-project-wrong](https://user-images.githubusercontent.com/636263/73774823-56c60400-478d-11ea-96d1-af598d511b80.jpg)

Maksims commented:
Fixed normals issue.

I believe this is ready now.

willeastcott commented:
This is great - we're so close to merging. Last outstanding thing for me:

If I merged this and deployed as v1.25.0, what will VR developers face? Especially those with PlayCanvas projects based on the old WebVR integration?

* Do we need a migration guide for existing PlayCanvas VR apps?
* What about VR controllers? What has to be done to re-enable those?

Maksims commented:
> If I merged this and deployed as v1.25.0, what will VR developers face? Especially those with PlayCanvas projects based on the old WebVR integration?

Currently this PR only enough for HMD (including mobile). WebVR is getting deprecated in most browsers, new Chrome wont even have its support behind the flag.

> * Do we need a migration guide for existing PlayCanvas VR apps?

We will need migration guide once we have all the functionality in place.

> * What about VR controllers? What has to be done to re-enable those?

Previously, WebVR had controllers implemented through GamePad API. WebXR implements own interfaces to interact with physical devices (even mouse and touch) and is done through XRInputSource.

**We can deprecate WebVR later, and deploy this, allowing developers to play with WebXR straight away. Later deprecate WebVR, as browsers slowly deprecate its APIs as well.**

Maksims commented:
Updated: WebVR is brought back. So we will deprecate it later.
And slightly simplified API.

willeastcott commented on src/framework/components/camera/component.js:
```diff
@@ -556,6 +568,51 @@ Object.assign(pc, function () {
+         * @param {pc.callbacks.XrError} [callback] - Optional callback function called once session is started. The callback has one argument Error - it is null if successfully started XR session.
+         * @example
+         * // On an entity with a camera component
+         * this.entity.camera.startXr(PC.XR_TYPE_IMMERSIVE_VR, function (err) {
```
PC -> pc

willeastcott commented on src/framework/components/camera/component.js:
```diff
@@ -556,6 +568,51 @@ Object.assign(pc, function () {
+         * @param {string} type - The type of session. Can be one of the following:
+         *
+         * * {@link pc.XR_TYPE_INLINE}: Inline - always available type of session. It has limited features availability and is rendered into HTML element.
+         * * {@link pc.XR_TYPE_IMMERSIVE_VR}: Immersive VR - session that provides exclusive access to VR device with best available tracking features.
```
In the interests of a simpler API, what about: `pc.XR_TYPE_IMMERSIVE_VR` -> `pc.XRTYPE_VR`? Same for AR. (BTW, the general naming convention is `pc.THISISTHETYPE_THISISTHESUBTYPE`, so one underscore. Yes, I know some enums break this).

willeastcott commented:
I definitely like how you've simplified some of the API new - nice move.

willeastcott commented on examples/xr/augmented-reality-basic.html:
```diff
@@ -0,0 +1,170 @@
+
+            if (app.xr.supported) {
+                var activate = function () {
+                    if (app.xr.isAvailable(pc.XR_TYPE_IMMERSIVE_AR)) {
```
So you have `app.xr.supported` and then `app.xr.isAvailable(pc.XR_TYPE_IMMERSIVE_AR)`. For consistency, and better readability, what about `app.xr.supports(pc.XRTYPE_AR)`

Maksims replied:
Support is static property, that is set before page is loaded. Which defines if platform supports WebXR API at all.
And "availability" - is what type of session currently is available, which can change during page life based on what devices are plugged in, etc.
I've split them intentionally, so not to confuse support with availability.

Support - cannot change, only browser upgrade can.
Availability - can change dynamically during page life.

Maksims commented:
Tested on Oculus Quest, works as intended.
